### PR TITLE
Implemented the Ctrl+C on the new Output Window

### LINKED ISSUE #7000: Ctrl + C when highlighting text on the Output Window doesn't work

### PR DESCRIPTION
**Pull request by N-thony**
Fixes #7000
@dannyparsons I have here implemented the Ctrl+C as you suggested in the corresponding issue with the Output Window.
@africanmathsinitiative/developers @rdstern this is ready for review. Thanks.
@ChrisMarsh82 I have additionally fixed the bug when copying text from `Output Window` to `Script Window` by Clicking on `Copy Tab` as I have explained in the issue. The change is made in `CopySelectedElementsToClipboard()`

### CODE REVIEW

**Comment by dannyparsons** [comment]
@ChrisMarsh82 Could you do a quick sanity check of this? It looks sensible to me.

@rdstern Could you also test?

File: instat/UserControl/ucrOutputPage.vb [hunk 1]
```diff
@@ -137,11 +137,14 @@ Public Class ucrOutputPage
...
-        dto.SetText(richText.Text, TextDataFormat.UnicodeText)
-        Clipboard.Clear()
-        Clipboard.SetDataObject(dto)
+        Dim strClip As String = String.Empty
+        With richText
+            For Each Line As String In .Lines
+                strClip &= Line & Environment.NewLine
+            Next
+            Clipboard.Clear()
+            Clipboard.SetText(strClip, TextDataFormat.Text)
```
Comment (dannyparsons) [inline instat/UserControl/ucrOutputPage.vb:146]: These lines shouldn't be duplicated. Extract to a Sub if it's required in multiple places.

Comment (N-thony) [reply instat/UserControl/ucrOutputPage.vb]: Thanks @dannyparsons, I resolved this.

**Review by rdstern — COMMENTED** [review]
I am not quite sure what is expected here.
I put some results into the output window.
I could then tick the box and perss the copy button, or ctrl C.  Then I could paste into the script window.  It pasted fine, but the font was changed, so it didn't look great.  I don't know if that is the copy/paste, or the script window?
I then tried to mark pare of the output in the output window to copy and didn't see how to do that.

**Comment by N-thony** [comment]
> I am not quite sure what is expected here. I put some results into the output window. I could then tick the box and perss the copy button, or ctrl C. Then I could paste into the script window. It pasted fine, but the font was changed, so it didn't look great. I don't know if that is the copy/paste, or the script window? I then tried to mark pare of the output in the output window to copy and didn't see how to do that.

Thanks @rdstern for your review and the discussion, I have made the change as suggested. For pasting the output from Describe>One Variable>Summarise dialogue into the script window, I think that may be an issue with the Script Window. I will have  look at it in a separate issue. Please take a look of the current change. Thanks.

**Review by rdstern — CHANGES_REQUESTED** [review]
@N-thony I can now mark a block in the output window.
But I now have some problems in the script window that were not there before.  I don't seem to be able to use ctrl V.  And more major is that the Clear does not seem to work any more in the script window.  
Pasting into the script window isn't the most important from the output window.  I am just using that because it is convenient.  Usually I will be pasting into word or Excel.  So I am not clear why anything needs to have changed in the script?

**Comment by N-thony** [comment]
> @N-thony I can now mark a block in the output window. But I now have some problems in the script window that were not there before. I don't seem to be able to use ctrl V. And more major is that the Clear does not seem to work any more in the script window. Pasting into the script window isn't the most important from the output window. I am just using that because it is convenient. Usually I will be pasting into word or Excel. So I am not clear why anything needs to have changed in the script?

@rdstern I was not able to reproduce the problems. Could you test again? Thanks.

**Comment by N-thony** [comment]
> > @N-thony I can now mark a block in the output window. But I now have some problems in the script window that were not there before. I don't seem to be able to use ctrl V. And more major is that the Clear does not seem to work any more in the script window. Pasting into the script window isn't the most important from the output window. I am just using that because it is convenient. Usually I will be pasting into word or Excel. So I am not clear why anything needs to have changed in the script?
> 
> @rdstern I was not able to reproduce the problems. Could you test again? Thanks.

@rdstern could you test this too? Thanks.

**Review by rdstern — APPROVED** [review]
Seems fine now

File: instat/UserControl/ucrOutputPage.vb [hunk 2]
```diff
@@ -270,6 +268,30 @@ Public Class ucrOutputPage
...
+                CopySelectedTextToClipBoard(richSelectedText)
+            Catch ex As Exception
+                MsgBox(ex.Message)
+            End Try
+        End If
+    End Sub
+
+    Private Sub CopySelectedTextToClipBoard(richText As RichTextBox)
+        Dim strClip As String = String.Empty
+        With richText
```
Comment (lloyddewit) [inline instat/UserControl/ucrOutputPage.vb:286]: What's the purpose of using 'With'? I undwerstand that, in this context, it just adds 2 extra lines of code and an extra indent.
Would it be more readable without?

File: instat/UserControl/ucrOutputPage.vb [hunk 3]
```diff
@@ -270,6 +268,30 @@ Public Class ucrOutputPage
...
+            For Each Line As String In .Lines
+                If Line <> "" Then strClip &= Line & Environment.NewLine
```
Comment (lloyddewit) [inline instat/UserControl/ucrOutputPage.vb:288]: What's the reason for not copying blank lines?
Should we comment the reason?

**Review by lloyddewit — CHANGES_REQUESTED** [review]
@N-thony Thank you, just a couple of open questions

File: instat/UserControl/ucrOutputPage.vb (hunk 3 above)
Comment (dannyparsons) [reply instat/UserControl/ucrOutputPage.vb]: Blank lines might be useful in a multi line copy, why ignore them?

**Comment by N-thony** [comment]
@lloyddewit thanks for your comments. I resolved them. Please take a look.

**Review by lloyddewit — DISMISSED** [review]

**Comment by lloyddewit** [comment]
@rdstern Since your last approval, there was just a small change that shouldn't have affected the functionality. Please could you approve again? thanks

**Review by dannyparsons — CHANGES_REQUESTED** [review]
In the current master if I copy some output using the Copy button in the Output Window e.g. from One Variable Summarise, then it pastes into Word with the formatting and colour.

When I try it on this PR the formatting is lost. So this needs to be retained.

**Comment by N-thony** [comment]
> In the current master if I copy some output using the Copy button in the Output Window e.g. from One Variable Summarise, then it pastes into Word with the formatting and colour.
> 
> When I try it on this PR the formatting is lost. So this needs to be retained.

@dannyparsons could you test this again? Thanks.

**Comment by dannyparsons** [comment]
Still doesn't work for me. Did you try this yourself and it worked?

**Comment by N-thony** [comment]
> Still doesn't work for me. Did you try this yourself and it worked?

@dannyparsons I did the test and it seems to work now.
image

**Comment by rdstern** [comment]
image

@N-thony Here for an example - from the review of copy first with marking copy and secondly with ctrl <C>

**Comment by N-thony** [comment]
> image
> 
> @N-thony Here for an example - from the review: First of copy having ticked a section and secondly marking the section and then using your ctrl C. I also tried all the different paste options in Word and can't get the same font/formatting as from the Copy button.

@rdstern thanks for your request. I have made a small change as shown below. Please take a look.

image

**Review by lloyddewit — APPROVED** [review]

**Comment by lloyddewit** [comment]
@dannyparsons I understand from @rdstern that you're happy now with this PR, and we can merge.
However, you requested a change earlier so I am blocked from merging.
Please could you approve and merge?
Thanks

**Review by dannyparsons — APPROVED** [review]

**Pull request merged**